lucene/core/src/java/org/apache/lucene/search/LRUQueryCache.java

```diff
@@ -732,8 +741,39 @@ public ScorerSupplier scorerSupplier(LeafReaderContext context) throws IOExcepti
 
       if (docIdSet == null) {
         if (policy.shouldCache(in.getQuery())) {
-          docIdSet = cache(context);
-          putIfAbsent(in.getQuery(), docIdSet, cacheHelper);
+          final ScorerSupplier supplier = in.scorerSupplier(context);
+          if (supplier == null) {
+            putIfAbsent(in.getQuery(), DocIdSet.EMPTY, cacheHelper);
+            return null;
+          }
+
+          final long cost = supplier.cost();
+          return new ScorerSupplier() {
+            @Override
+            public Scorer get(long leadCost) throws IOException {
+              // skip cache operation which would slow query down too much
+              if ((cost > skipCacheCost || cost > leadCost * skipCacheFactor)
```

Comment:
We have tested different scenarios to observe the query latency with/without cacheing in an online ES cluster. Here is the result:

| queryPattern      | latencyWithoutCaching  | latencyWithCaching | leadCost | rangeQueryCost  | skipCacheFactor |
| ---------- | :-----------:  | :-----------: | :-----------:  | :-----------: | :-----------:  |
| ip:xxx AND time:[t-1h, t] | 10ms | 36ms(+260%) | 20528 | 878979 | 42 |
| ip:xxx AND time:[t-4h, t] | 10ms | 100ms(+900%) | 20528 | 4365870 | 212 |
| ip:xxx AND time:[t-8h, t] | 11ms | 200ms(+1700%) | 20528 | 8724483 | 425 |
| ip:xxx AND time:[t-12h, t] | 12ms | 300ms(+2400%) | 20528 | 13083096 | 637 |
| ip:xxx AND time:[t-24h, t] | 16ms | 500ms(+3000%) | 20528 | 26158936 | 1274 |
| ip:xxx AND time:[t-48h, t] | 30ms | 1200ms(3900%) | 20528 | 52310616 | 2548 |

As the table shows, query latency without caching is low and it's related with the final result set. Query latency with caching is much high and it's mainly related with _rangeQueryCost_. According to the above test, we set the default value of _skipCacheFactor_ to 250, which make the query slower by no more than 10 times.

In addition to _skipCacheFactor_ which is similar to _maxCostFactor_ in LUCENE-8027, we add a new parameter _skipCacheCost_. The mainly reasons are:
- control the time used for caching as the caching time is related to the cost of range query.
- skip caching too large range queries which will consume too much memory and evict cache frequently.

How do you think? Looking forward to your ideas. @jpountz

Reply:
> control the time used for caching as the caching time is related to the cost of range query

Understood, but if the query has a 250x higher cost already, then it's probably fine right?

> skip caching too large range queries which will consume too much memory and evict cache entries frequently

Maybe a more intuitive way to do that would be to call this parameter `maxRamBytesUsedPerEntry` and skip caching if `min(maxDoc/Byte.SIZE /* assuming bitset */, cost * Short.BYTES /* assuming short[] */) > maxRamBytesUsedPerEntry`?

Reply:
Yes. With a 250x higher cost, it’s good enough now without _skipCacheCost_.

Also we find that _skipCacheCost_ could be done through the following code. Although we think that the allowed cache entry size is a little large, we are not going to modify it now. If we find that the demand is very strong, we will open a new PR.
https://github.com/apache/lucene-solr/blob/61a00822e78a02f14ac060dabf66c2b5053c12d7/lucene/core/src/java/org/apache/lucene/search/LRUQueryCache.java#L679-L691